### PR TITLE
LCOW: Add UVM debugability by grabbing logs before tear-down

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@rn @johnstep - Adds the ability for docker to grab logs from the utility VM before it is torn down in advanced debugging mode (the daemon must be in debug mode itself, plus "OPENGCS_DEBUG_ENABLE" must be set in its environment.)

You can see an example of what it looks like from https://github.com/Microsoft/opengcs/pull/134.

It's a far-from-perfect solution, but pretty much the best that is possible in the RS3 release.

### CODE REVIEW

**Comment by jhowardmsft** [comment]
@friism FYI

File: libcontainerd/client_windows.go [hunk 1]
```diff
@@ -439,6 +439,11 @@ func (clnt *client) AddProcess(ctx context.Context, containerID, processFriendly
 	if err != nil {
 		return -1, err
 	}
+
+	defer func() {
```
Comment (simonferquel) [inline libcontainerd/client_windows.go]: could be `defer container.debugGCS()`

File: libcontainerd/container_windows.go [hunk 2]
```diff
@@ -58,6 +59,10 @@ func (ctr *container) start(attachStdio StdioCallback) error {
 		return err
 	}
 
+	defer func() {
+		ctr.debugGCS()
```
Comment (simonferquel) [inline libcontainerd/container_windows.go]: could be `defer ctr.debugGCS()`

**Review by simonferquel — CHANGES_REQUESTED** [review]

File: libcontainerd/client_windows.go (hunk 1 above)
Comment (jhowardmsft) [reply libcontainerd/client_windows.go]: Updated

File: libcontainerd/container_windows.go (hunk 2 above)
Comment (jhowardmsft) [reply libcontainerd/container_windows.go]: Updated

File: libcontainerd/container_windows.go [hunk 3]
```diff
@@ -58,6 +59,8 @@ func (ctr *container) start(attachStdio StdioCallback) error {
 		return err
 	}
 
+	defer ctr.debugGCS()
+
```
Comment (rn) [inline libcontainerd/container_windows.go]: Given that `defer ctr.debugGCS()` is called both on the error path and the normal path, wouldn't it be cleaner to write this (and the above) as:
```
err = ctr.hcsContainer.Start()
defer ctr.debugGCS()
if err != nil {
...
```

**Review by rn — COMMENTED** [review]
Just a minor style nit but otherwise LGTM. I've used this yesterday and it was very helpful for debugging indeed.

File: libcontainerd/container_windows.go (hunk 3 above)
Comment (jhowardmsft) [reply libcontainerd/container_windows.go]: Ah no. If I deferred the one after start, it wouldn't fire as terminate would have been called by the time the defer was executed, and there wouldn't be a UVM around to debug. Make sense?

Comment (rn) [reply libcontainerd/container_windows.go]: yes, sorry, missed the subtlety.

**Review by rn — APPROVED** [review]
Not a maintainer, but this LGTM and I've used it for debug last week and it was very useful

File: libcontainerd/utils_windows.go [hunk 4]
```diff
@@ -19,3 +23,16 @@ func setupEnvironmentVariables(a []string) map[string]string {
 func (s *LCOWOption) Apply(interface{}) error {
 	return nil
 }
+
+// DebugGCS is a dirty hack for debugging for Linux Utility VMs. It simply
```
Comment (thaJeztah) [inline libcontainerd/utils_windows.go]: s/DebugGCS/debugGCS/

**Review by thaJeztah — APPROVED** [review]
LGTM ping @simonferquel PTAL

File: libcontainerd/utils_windows.go (hunk 4 above)
Comment (jhowardmsft) [reply libcontainerd/utils_windows.go]: Fixed

**Comment by thaJeztah** [comment]
vendor CI is failing;" :cry:

```
05:05:23 The result of vndr differs
05:05:23 
05:05:23  D vendor/github.com/gotestyourself/gotestyourself/skip/skip.go
05:05:23 
05:05:23 Please vendor your package with github.com/LK4D4/vndr.
05:05:23

**Comment by jhowardmsft** [comment]
Weird! Prob won't get time to do today though

**Comment by jhowardmsft** [comment]
Truly bizarre. vndr for me showed no difference - both a full run and a specific vndr of `gotestyourself/gotestyourself v1.1.0`. So manually fixed up....

```
jhoward@jhoward-z420:[~/go/src/github.com/docker/docker]: $ grep gotestyourself vendor.conf
github.com/gotestyourself/gotestyourself v1.1.0
jhoward@jhoward-z420:[~/go/src/github.com/docker/docker]: $ vndr github.com/gotestyourself/gotestyourself v1.1.0
2017/09/18 09:27:10 Collecting initial packages
2017/09/18 09:27:11 Download dependencies
2017/09/18 09:27:13     Clone github.com/gotestyourself/gotestyourself, revision v1.1.0
2017/09/18 09:27:14     Finished clone github.com/gotestyourself/gotestyourself
2017/09/18 09:27:14 Dependencies downloaded. Download time: 1.0286058s
2017/09/18 09:27:14 Collecting all dependencies
2017/09/18 09:27:24 Clean vendor dir from unused packages
2017/09/18 09:27:24 Success
2017/09/18 09:27:24 Running time: 13.7501532s
jhoward@jhoward-z420:[~/go/src/github.com/docker/docker]: $ git status
On branch jjh/debuggcs
Your branch is up-to-date with 'msdocker/jjh/debuggcs'.
nothing to commit, working directory clean
jhoward@jhoward-z420:[~/go/src/github.com/docker/docker]: $
```

**Comment by jhowardmsft** [comment]
CI vendoring passed. Removing failing CI label :)

**Comment by jhowardmsft** [comment]
The janky failure doesn't look related - seems to be failing on multiple PRs.

If I'm reading it right, it's docker-py tests which are failing. @shin any ideas?

```
02:09:28.668  generated xml file: /go/src/github.com/docker/docker/bundles/test-docker-py/results.xml 
02:09:28.669 =========================== short test summary info ============================
02:09:28.669 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_init_path
02:09:28.669   init-path removed in 17.05.0
02:09:28.669 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_storage_opt
02:09:28.669   Not supported on most drivers
02:09:28.670 XFAIL ../../../../../docker-py/tests/integration/models_images_test.py::ImageCollectionTest::test_build_with_error
02:09:28.670   Engine 1.13 responds with status 500
02:09:28.670 SKIP [1] tests/integration/api_image_test.py:270: Doesn't work inside a container - FIXME
02:09:28.670 SKIP [1] tests/integration/api_service_test.py:287: API version is too low (< 1.30)
02:09:28.670 SKIP [1] tests/integration/api_network_test.py:454: API version is too low (< 1.29)
02:09:28.670 SKIP [1] /docker-py/tests/helpers.py:66: Feature requires Docker Engine experimental mode
02:09:28.671 SKIP [1] tests/integration/api_healthcheck_test.py:51: API version is too low (< 1.29)
02:09:28.671 SKIP [1] tests/integration/api_build_test.py:191: API version is too low (< 1.29)
02:09:28.671 SKIP [1] tests/integration/api_service_test.py:300: API version is too low (< 1.27)
02:09:28.671 SKIP [1] /docker-py/tests/integration/api_swarm_test.py:24: Test stalls the engine on 1.12.0
02:09:28.671 SKIP [1] tests/integration/models_services_test.py:84: Makes Swarm unstable?
02:09:28.671 ============== 244 passed, 9 skipped, 3 xfailed in 313.20 seconds ==============
02:09:28.700 ---> Making bundle: .integration-daemon-stop (in bundles/test-docker-py)
```

**Comment by jhowardmsft** [comment]
Sorry @shin - I meant @shin-. ^^

**Comment by thaJeztah** [comment]
Wonder if there's a bug in vndr; there was an issue on one of the other LCOW PR's as well with the gotestyourself vendoring

**Comment by jhowardmsft** [comment]
(Already mentions on slack, but I suspect it's a bug in WSL)

**Comment by jhowardmsft** [comment]
ping @simonferquel PTAL. Thanks!

**Review by simonferquel — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
😞 looks like we now have a missing dependency, but vendor check says it's ok?

```
09:10:29 integration/service/inspect_test.go:22:2:warning: undeclared name: skip (gosimple)
09:10:29 integration/service/inspect_test.go:15:2:warning: could not import github.com/gotestyourself/gotestyourself/skip (cannot find package "github.com/gotestyourself/gotestyourself/skip" in any of: (gosimple)
```

@dnephin vndr bug?

**Comment by dnephin** [comment]
What's even more bizarre is that the powerpc and z builds actually PASS, even though the log claims that this package was removed. I see it running the `TestInspect` which uses this `skip()`. So something is very broken on those builds. 

I'm not sure what's up with `vndr` but it works for everyone else on linux.

**Comment by jhowardmsft** [comment]
I installed an Ubuntu 17.04 instance and ran vndr from there. Let's see what happens this time first...

**Comment by dnephin** [comment]
>  but vendor check says it's ok?

Right that's concerning too. I'll look into the old vendor job to see if the log has anything

**Comment by jhowardmsft** [comment]
That seemed OK, all jobs got to start running integration tests and vendor passed. So squashed the revendor commits and pushed update. Fingers crossed.

**Comment by jhowardmsft** [comment]
Yay, green!!!!

**Comment by thaJeztah** [comment]
whooooop!

**Pull request merged**